src/main/java/seedu/mark/model/Model.java

```diff
@@ -84,4 +84,29 @@
      * @throws NullPointerException if {@code predicate} is null.
      */
     void updateFilteredBookmarkList(Predicate<Bookmark> predicate);
+
+    /**
+     * Returns true if the model has previous mark states to restore.
+     */
+    boolean canUndoMark();
+
+    /**
+     * Returns true if the model has undone mark states to restore.
+     */
+    boolean canRedoMark();
+
+    /**
+     * Restores the model's mark to its previous state.
+     */
+    void undoMark();
+
+    /**
+     * Restores the model's mark to its previously undone state.
+     */
+    void redoMark();
+
+    /**
+     * Saves the current mark state for undo/redo.
+     */
+    void commitMark();
```

Comment:
Can we rename this? (and other usages of `commitMark()`)
```suggestion
    void saveMark();
```

Reply:
Okay!